README.md

```diff
@@ -32,7 +28,7 @@ which makes it very easy to get started and also helps to write your own `Qlever
 QLever via docker (which is the default setting), the script pulls the most recent docker image automatically and you
 don't have to download or compile the code.
 
-# Advanced feature, in-depth information, and older documentation
+If the `qlever` script does not work for you for whatever reason, have a look at the [Dockerfile for Ubuntu 22.04](https://github.com/ad-freiburg/qlever/blob/master/Dockerfile) or the [Dockerfiles for older Ubuntu versions](https://github.com/ad-freiburg/qlever/tree/master/Dockerfiles). The [source code of the qlever script](https://github.com/ad-freiburg/qlever-control/blob/main/qlever) also provide information how to use QLever (in particular, note the functions `action_start` and `action_index`).
```

Comment:
Hey, thanks for these README updates from a new user :)

I'm working on making qlever work better on MacOS, and the `qlever-control` script doesn't work at all (because MacOS uses zsh, and its version of bash and many other core commands like `sed` have slightly different arguments and effects because they're BSD based instead of GNU). I was considering porting [parts of] the qlever-control script to a portable language like python, but I could also skip that and just focus on better documenting how to run qlever directly on MacOS, with more examples. What would you recommend?

Reply:
@bradenmacdonald Yes, we are aware of that and we would love to make it work for MacOS and other platforms, too. I have already pondered the question quite a bit what the right language for such a control program is.

The advantage of a bash-like script is that you can write commands directly in the configuration file, for example, see https://github.com/ad-freiburg/qlever-control/blob/main/Qleverfiles/Qleverfile.wikidata or the other Qleverfiles in that directory. It's so nice when you have a single configuration file where you can write expressive stuff easily.

The disadvantage is that it's hard to make this portable. I also considered writing the control script in Python, but I wasn't sure how well executing all sorts of programs from Python works in a portable way. Do you have any experience with this?

Another aspect was that I wanted to have command-line autocompletion in the shell. With a bash script this comes naturally, but there is again the issue with other platforms.

Reply:
Thanks for the additional context. I'm afraid I don't have a good answer, as it is a tough question.

The examples you linked to are a nice illustration of how powerful and concise the current approach is, though I also think most if not all of the commands in the Wikidata file should be portable enough to work on both MacOS and Linux, or could easily be made portable. Generally there is some subset of the POSIX commands that are reasonably similar on both platforms.

One approach I lean toward would be to make the `qlever` script itself a python script, which would not only make it portable but I think also make it easier to read and maintain, then to use it to run shell commands for the actual Qleverfile scripts, like the examples you showed me, and to do things like enabling bash completion.

That would require that completions are installed a bit differently; instead of sourcing the file, you'd run it and pipe the result into your shell rc file, but I think that's a more customary way to do this anyways and has the benefit of being permanently set up on your system. See https://deno.com/manual@v1.3.3/getting_started/setup_your_environment#shell-autocomplete for an example that I'm familiar with.

Another approach is to run everything in Docker in a very lightweight container with a consistent shell environment, and have the `qlever-indices` and `qlever-code` as volumes mounted within the shell container workspace. But the problem is that if qlever is also spawned within the same container or a peer container, then accessing files on the host via mounted volumes is quite slow on MacOS and would have poor performance for things like index building. If qlever were running as a server on the host, and the lightweight shell container only interacted with it via network, it would be fine, but I don't think that setup is feasible given how qlever works for things like index building.

Reply:
Contonuing this on https://github.com/ad-freiburg/qlever-control/issues/16 (this important discussion mertis an own issue) and merging this PR now.